File: rust/crates/h2ts-server/src/handshake.rs
//! Server-side WebSocket handshake (RFC 6455).
//!
//! It's small: parse the upgrade headers, compute `Sec-WebSocket-Accept` (SHA-1
//! of the client key plus the RFC 6455 magic GUID, base64-encoded), reply `101`,
//! and hand back hyper's upgraded connection as a raw byte stream. The framing on
//! top of that stream is done by [`bridge`](crate::bridge) using wslay.

use std::fmt;
use std::future::Future;

use base64::Engine as _;
use bytes::Bytes;
use http::header::{
    HeaderName, HeaderValue, CONNECTION, SEC_WEBSOCKET_ACCEPT, SEC_WEBSOCKET_KEY,
    SEC_WEBSOCKET_PROTOCOL, UPGRADE,
};
use http_body_util::Empty;
use hyper::upgrade::Upgraded;
use hyper::{Request, Response, StatusCode};
use hyper_util::rt::TokioIo;
use sha1::{Digest, Sha1};

/// RFC 6455 §4.2.2: the magic GUID appended to the client key before hashing.
const WS_GUID: &[u8] = b"258EAFA5-E914-47DA-95CA-C5AB0DC85B11";

/// The subprotocol h2ts clients offer by default. [`accept`] echoes it when the
/// client offered it and the handler didn't choose another (see [`accept_with`]).
pub const DEFAULT_SUBPROTOCOL: &str = "h2ts";

/// Options controlling how the handshake picks a subprotocol to echo when the
/// handler's selector declines (see [`accept_with_options`]).
#[derive(Debug, Clone, Copy, Default)]
pub struct AcceptOptions {
    /// When the selector returns `None` and the client did **not** offer
    /// [`DEFAULT_SUBPROTOCOL`] (`h2ts`), accept the client's first offered
    /// subprotocol (or, if it offered none, complete with no subprotocol)
    /// instead of **rejecting** the handshake.
    ///
    /// Off by default: a client that speaks neither `h2ts` nor a
    /// handler-selected protocol is rejected with
    /// [`WebSocketError::UnsupportedSubprotocol`]. Turn this on for a
    /// codec-agnostic tunnel that accepts whatever framing the client offered
    /// (e.g. a raw/binary tunnel, or a websockify-style `binary` client).
    pub allow_implicit_codec: bool,
}

/// The upgraded WebSocket connection, presented as a raw tokio byte stream
/// (`AsyncRead + AsyncWrite`). Hand it to [`bridge`](crate::bridge),
/// [`serve_h2`](crate::serve_h2), or [`WsByteStream`](crate::WsByteStream).
pub type UpgradedIo = TokioIo<Upgraded>;

/// Errors from the WebSocket handshake.
#[derive(Debug)]
pub enum WebSocketError {
    /// The request was not a valid WebSocket upgrade (missing/invalid
    /// `Upgrade` / `Connection` / `Sec-WebSocket-Key`).
    NotUpgradeRequest,
    /// The client offered no subprotocol the server accepts: it didn't offer
    /// [`DEFAULT_SUBPROTOCOL`] (`h2ts`), the handler's selector declined, and
    /// [`AcceptOptions::allow_implicit_codec`] was off. Reject it with a `400`
    /// (see [`WebSocketError::rejection_response`]).
    UnsupportedSubprotocol,
    /// hyper failed to upgrade the connection after the `101` was sent.
    Upgrade(hyper::Error),
}

impl WebSocketError {
    /// The HTTP response to reject this handshake with, for the pre-upgrade
    /// errors returned by [`accept`] / [`accept_with`] / [`accept_with_options`]:
    /// `426 Upgrade Required` for a non-WebSocket request, `400 Bad Request` for
    /// an unsupported subprotocol. (An [`Upgrade`](WebSocketError::Upgrade) error
    /// happens *after* the `101` and has no meaningful rejection response; it maps
    /// to `500` for completeness.)
    ///
    /// ```ignore
    /// let (response, ws_fut) = match h2ts_server::accept(&mut req) {
    ///     Ok(pair) => pair,
    ///     Err(err) => return Ok(err.rejection_response()), // send the 4xx back
    /// };
    /// ```
    pub fn rejection_response(&self) -> Response<Empty<Bytes>> {
        let status = match self {
            WebSocketError::NotUpgradeRequest => StatusCode::UPGRADE_REQUIRED,
            WebSocketError::UnsupportedSubprotocol => StatusCode::BAD_REQUEST,
            WebSocketError::Upgrade(_) => StatusCode::INTERNAL_SERVER_ERROR,
        };
        Response::builder()
            .status(status)
            .body(Empty::new())
            .expect("static rejection response is well-formed")
    }
}

impl fmt::Display for WebSocketError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            WebSocketError::NotUpgradeRequest => f.write_str("not a WebSocket upgrade request"),
            WebSocketError::UnsupportedSubprotocol => {
                f.write_str("client offered no supported subprotocol")
            }
            WebSocketError::Upgrade(e) => write!(f, "WebSocket upgrade failed: {e}"),
        }
    }
}

impl std::error::Error for WebSocketError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            WebSocketError::Upgrade(e) => Some(e),
            WebSocketError::NotUpgradeRequest | WebSocketError::UnsupportedSubprotocol => None,
        }
    }
}

/// Whether `header` is present and lists `needle` as one of its comma-separated,
/// case-insensitive tokens.
fn header_lists<B>(request: &Request<B>, header: HeaderName, needle: &str) -> bool {
    request
        .headers()
        .get(header)
        .and_then(|v| v.to_str().ok())
        .map(|list| {
            list.split(',')
                .any(|t| t.trim().eq_ignore_ascii_case(needle))
        })
        .unwrap_or(false)
}

/// Whether `request` is a WebSocket upgrade: `Upgrade: websocket`,
/// `Connection: upgrade`, and a `Sec-WebSocket-Key`.
pub fn is_upgrade_request<B>(request: &Request<B>) -> bool {
    header_lists(request, UPGRADE, "websocket")
        && header_lists(request, CONNECTION, "upgrade")
        && request.headers().contains_key(SEC_WEBSOCKET_KEY)
}

/// The WebSocket subprotocols the client offered, in order, whitespace-trimmed;
/// empty if it offered none. Gives a handler full visibility into the offer so
/// it can pass one to [`accept_with`].
pub fn offered_protocols<B>(request: &Request<B>) -> Vec<&str> {
    request
        .headers()
        .get(SEC_WEBSOCKET_PROTOCOL)
        .and_then(|v| v.to_str().ok())
        .map(|list| {
            list.split(',')
                .map(str::trim)
                .filter(|s| !s.is_empty())
                .collect()
        })
        .unwrap_or_default()
}

/// The outcome of the default subprotocol fallback, when the handler's selector
/// declined.
#[derive(Debug, PartialEq, Eq)]
enum Fallback {
    /// Complete the handshake, echoing this subprotocol (`Some`) or none (`None`).
    Accept(Option<String>),
    /// Reject the handshake: the client offered no acceptable subprotocol and
    /// implicit codecs are disabled.
    Reject,
}

/// The subprotocol decision when the handler's selector declined.
///
/// Prefers [`DEFAULT_SUBPROTOCOL`] (`h2ts`) whenever the client offered it (in
/// its exact offered casing). Otherwise, with `allow_implicit_codec`, accept the
/// client's first offered subprotocol (or none, if it offered nothing); without
/// it, [`Fallback::Reject`] — we don't commit to a codec we don't understand.
fn fallback_subprotocol(offered: &[&str], allow_implicit_codec: bool) -> Fallback {
    if let Some(p) = offered
        .iter()
        .find(|p| p.eq_ignore_ascii_case(DEFAULT_SUBPROTOCOL))
    {
        return Fallback::Accept(Some(p.to_string()));
    }
    if allow_implicit_codec {
        return Fallback::Accept(offered.first().map(|p| p.to_string()));
    }
    Fallback::Reject
}

/// Compute the `Sec-WebSocket-Accept` value for a client `Sec-WebSocket-Key`.
fn accept_key(key: &[u8]) -> String {
    let mut hasher = Sha1::new();
    hasher.update(key);
    hasher.update(WS_GUID);
    base64::engine::general_purpose::STANDARD.encode(hasher.finalize())
}

/// Accept a WebSocket upgrade, choosing which offered subprotocol to echo, with
/// explicit [`AcceptOptions`].
///
/// `select` is handed the full list of subprotocols the client offered (the same
/// thing [`offered_protocols`] returns) and returns the one to echo in the `101`,
/// or `None` to decline. When it declines, [`DEFAULT_SUBPROTOCOL`] (`h2ts`) is
/// echoed if the client offered it; otherwise the handshake is **rejected** with
/// [`WebSocketError::UnsupportedSubprotocol`] — unless
/// [`AcceptOptions::allow_implicit_codec`] is set, which instead accepts the
/// client's first offered codec (or completes with none if it offered nothing).
/// Turn a rejection into a `4xx` to send back via
/// [`WebSocketError::rejection_response`].
///
/// Per RFC 6455 a client that offered subprotocols will *fail* the connection if
/// the server echoes one it did not offer, so `select` should return a member of
/// the offered list (or `None`).
///
/// See [`accept_with`] (default options) and [`accept`] (the common case).
// The returned `impl Future` can't be aliased away, so the tuple reads as
// "complex"; it's just (response, upgrade-future).
#[allow(clippy::type_complexity)]
pub fn accept_with_options<B, F>(
    request: &mut Request<B>,
    select: F,
    options: AcceptOptions,
) -> Result<
    (
        Response<Empty<Bytes>>,
        impl Future<Output = Result<UpgradedIo, WebSocketError>>,
    ),
    WebSocketError,
>
where
    F: FnOnce(&[&str]) -> Option<String>,
{
    if !is_upgrade_request(request) {
        return Err(WebSocketError::NotUpgradeRequest);
    }
    // Present because `is_upgrade_request` checked for it.
    let key = request
        .headers()
        .get(SEC_WEBSOCKET_KEY)
        .ok_or(WebSocketError::NotUpgradeRequest)?;
    let accept_value = accept_key(key.as_bytes());

    // Let the handler pick from the offered list; otherwise apply the default
    // fallback (h2ts if offered, then optionally the first offered codec, else
    // reject).
    let offered = offered_protocols(request);
    let decision = match select(&offered) {
        // Never echo a subprotocol the client did not offer (RFC 6455 §4.2.2 — the
        // client fails the connection if we do). A selection outside the offered
        // list is treated as a decline and falls back to the default policy.
        Some(proto) if offered.iter().any(|o| o.eq_ignore_ascii_case(&proto)) => {
            Fallback::Accept(Some(proto))
        }
        _ => fallback_subprotocol(&offered, options.allow_implicit_codec),
    };
    drop(offered); // release the immutable borrow before upgrading below
    let chosen = match decision {
        Fallback::Accept(proto) => proto,
        Fallback::Reject => return Err(WebSocketError::UnsupportedSubprotocol),
    };

    // base64 output is always valid header-value ASCII, so this never fails.
    let accept_header = HeaderValue::from_str(&accept_value).expect("base64 is valid header ASCII");
    let mut response = Response::builder()
        .status(StatusCode::SWITCHING_PROTOCOLS)
        .header(CONNECTION, HeaderValue::from_static("Upgrade"))
        .header(UPGRADE, HeaderValue::from_static("websocket"))
        .header(SEC_WEBSOCKET_ACCEPT, accept_header)
        .body(Empty::<Bytes>::new())
        .expect("static 101 response is well-formed");
    if let Some(proto) = chosen {
        // Skip a subprotocol that isn't a valid header value rather than fail.
        if let Ok(value) = HeaderValue::from_str(&proto) {
            response.headers_mut().insert(SEC_WEBSOCKET_PROTOCOL, value);
        }
    }

    // Capture the upgrade future now; it resolves once the 101 is flushed and
    // hyper hands over the connection.
    let on_upgrade = hyper::upgrade::on(&mut *request);
    let fut = async move {
        let upgraded = on_upgrade.await.map_err(WebSocketError::Upgrade)?;
        Ok(TokioIo::new(upgraded))
    };
    Ok((response, fut))
}

/// Accept a WebSocket upgrade, choosing which offered subprotocol to echo.
///
/// Like [`accept_with_options`] with the default [`AcceptOptions`]: when `select`
/// declines, [`DEFAULT_SUBPROTOCOL`] (`h2ts`) is echoed if the client offered it;
/// otherwise the handshake is **rejected**
/// ([`WebSocketError::UnsupportedSubprotocol`]). Use [`accept_with_options`] with
/// `allow_implicit_codec` to instead accept the client's first offered codec.
///
/// ```ignore
/// // Prefer "chat" if the client offered it, otherwise fall back to h2ts.
/// let (response, ws_fut) = h2ts_server::accept_with(&mut req, |offered| {
///     offered.iter().find(|p| p.eq_ignore_ascii_case("chat")).map(|p| p.to_string())
/// })?;
/// ```
#[allow(clippy::type_complexity)]
pub fn accept_with<B, F>(
    request: &mut Request<B>,
    select: F,
) -> Result<
    (
        Response<Empty<Bytes>>,
        impl Future<Output = Result<UpgradedIo, WebSocketError>>,
    ),
    WebSocketError,
>
where
    F: FnOnce(&[&str]) -> Option<String>,
{
    accept_with_options(request, select, AcceptOptions::default())
}

/// Accept a WebSocket upgrade, requiring the [`DEFAULT_SUBPROTOCOL`] (`h2ts`).
///
/// Echoes `h2ts` when the client offered it; a client that doesn't offer `h2ts`
/// is **rejected** with [`WebSocketError::UnsupportedSubprotocol`] (send
/// [`rejection_response`](WebSocketError::rejection_response) back as a `400`).
/// Use [`accept_with`] to select a different offered subprotocol, or
/// [`accept_with_options`] with `allow_implicit_codec` to accept any codec.
///
/// On success, returns the `101 Switching Protocols` response to send back
/// immediately, plus a future that resolves to the upgraded connection as a byte
/// stream ([`UpgradedIo`]). Drive the response through your framework; spawn the
/// future and hand the stream to [`bridge`](crate::bridge) or
/// [`serve_h2`](crate::serve_h2). Control frames (ping/close) are handled
/// downstream by wslay.
///
/// The outer HTTP/1 connection must be served with upgrades enabled
/// (`http1::Builder::serve_connection(..).with_upgrades()`).
#[allow(clippy::type_complexity)]
pub fn accept<B>(
    request: &mut Request<B>,
) -> Result<
    (
        Response<Empty<Bytes>>,
        impl Future<Output = Result<UpgradedIo, WebSocketError>>,
    ),
    WebSocketError,
> {
    accept_with(request, |_offered| None)
}

File: go/server/handshake.go
// Server-side WebSocket handshake (RFC 6455).
//
// It's small: validate the upgrade headers, pick which offered subprotocol to
// echo (see the negotiation policy on [AcceptOptions]), compute
// Sec-WebSocket-Accept (SHA-1 of the client key plus the RFC 6455 magic GUID,
// base64-encoded), hijack the connection, and write the 101. The framing on top
// of the returned byte stream is done by [Conn].

package server

import (
	"bufio"
	"crypto/sha1"
	"encoding/base64"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// wsGUID is the RFC 6455 §4.2.2 magic GUID appended to the client key before
// hashing to derive Sec-WebSocket-Accept.
const wsGUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

// AcceptOptions controls how [AcceptWithOptions] negotiates a subprotocol and
// which origins it accepts.
type AcceptOptions struct {
	// Select is handed the full list of subprotocols the client offered (the
	// same thing [OfferedProtocols] returns) and returns the one to echo in the
	// 101, or "" to decline. A selection outside the offered list is treated as a
	// decline (RFC 6455 forbids echoing a subprotocol the client did not offer).
	// When it declines, the default policy applies: echo DefaultSubprotocol
	// (h2ts) if the client offered it, else reject (or, with AllowImplicitCodec,
	// echo the client's first offered codec). nil means always decline.
	Select func(offered []string) string

	// AllowImplicitCodec, when the selector declines and the client did not offer
	// h2ts, accepts the client's first offered subprotocol (or completes with no
	// subprotocol if it offered none) instead of rejecting the handshake. Off by
	// default: a client speaking neither h2ts nor a selected protocol is rejected
	// with 400. Turn it on for a codec-agnostic tunnel (a websockify-style
	// `binary` client, a raw tunnel, …).
	AllowImplicitCodec bool

	// AllowedOrigins is an opt-in Origin allowlist (Cross-Site WebSocket Hijacking
	// defence). When non-nil, only handshakes whose Origin header matches one of
	// these entries (ASCII case-insensitive, e.g. "https://app.example.com") are
	// accepted; everything else — including a missing Origin — is rejected with
	// 403. nil (the default) accepts any origin, mirroring nginx (which does no
	// Origin validation by default); a browser always sends Origin, so a
	// legitimate browser client is unaffected.
	AllowedOrigins []string
}

// HandshakeError is a pre-upgrade handshake rejection. [Accept] /
// [AcceptWithOptions] write the response with [HandshakeError.Status] to the
// [http.ResponseWriter] before returning it, so a handler can simply return; the
// error is exposed for logging and inspection.
type HandshakeError struct {
	// Status is the HTTP status the handshake was rejected with: 426 (not a
	// WebSocket upgrade), 400 (no supported subprotocol), or 403 (forbidden
	// origin).
	Status int
	// Reason is a short human-readable description.
	Reason string
}

func (e *HandshakeError) Error() string {
	return fmt.Sprintf("h2ts handshake rejected: %s (HTTP %d)", e.Reason, e.Status)
}

// The pre-upgrade rejections, mirroring the Rust server's WebSocketError arms.
var (
	errNotUpgrade             = &HandshakeError{http.StatusUpgradeRequired, "not a WebSocket upgrade request"}
	errUnsupportedSubprotocol = &HandshakeError{http.StatusBadRequest, "client offered no supported subprotocol"}
	errForbiddenOrigin        = &HandshakeError{http.StatusForbidden, "origin not allowed"}
)

// headerListContains reports whether any value of header (which may be a
// comma-separated token list) contains needle, ASCII case-insensitively.
func headerListContains(h http.Header, header, needle string) bool {
	for _, v := range h.Values(header) {
		for _, tok := range strings.Split(v, ",") {
			if strings.EqualFold(strings.TrimSpace(tok), needle) {
				return true
			}
		}
	}
	return false
}

// IsUpgradeRequest reports whether r is a WebSocket upgrade request: Upgrade
// lists websocket, Connection lists upgrade, and a Sec-WebSocket-Key is present.
func IsUpgradeRequest(r *http.Request) bool {
	return headerListContains(r.Header, "Upgrade", "websocket") &&
		headerListContains(r.Header, "Connection", "upgrade") &&
		r.Header.Get("Sec-WebSocket-Key") != ""
}

// OfferedProtocols returns the WebSocket subprotocols the client offered, in
// order, whitespace-trimmed and with empties dropped; nil if it offered none.
// Handles both a single comma-separated header and repeated headers.
func OfferedProtocols(r *http.Request) []string {
	var out []string
	for _, v := range r.Header.Values("Sec-WebSocket-Protocol") {
		for _, tok := range strings.Split(v, ",") {
			if tok = strings.TrimSpace(tok); tok != "" {
				out = append(out, tok)
			}
		}
	}
	return out
}

// containsFold reports whether list holds s, ASCII case-insensitively.
func containsFold(list []string, s string) bool {
	for _, v := range list {
		if strings.EqualFold(v, s) {
			return true
		}
	}
	return false
}

// negotiate decides which subprotocol to echo. It returns (chosen, hasChosen,
// accept): accept=false means reject with 400; hasChosen=false with accept=true
// means complete the handshake echoing no subprotocol.
//
// Prefers DefaultSubprotocol (h2ts) — in its exact offered casing — whenever the
// client offered it. Otherwise, with allowImplicit, echoes the client's first
// offered codec (or none); without it, rejects.
func negotiate(offered []string, sel func([]string) string, allowImplicit bool) (chosen string, hasChosen, accept bool) {
	// Never echo a subprotocol the client did not offer (RFC 6455 §4.2.2 — the
	// client fails the connection if we do); an unoffered selection is a decline.
	if sel != nil {
		if s := sel(offered); s != "" && containsFold(offered, s) {
			return s, true, true
		}
	}
	for _, p := range offered {
		if strings.EqualFold(p, DefaultSubprotocol) {
			return p, true, true
		}
	}
	if allowImplicit {
		if len(offered) > 0 {
			return offered[0], true, true
		}
		return "", false, true
	}
	return "", false, false
}

// acceptKey computes the Sec-WebSocket-Accept value for a client
// Sec-WebSocket-Key.
func acceptKey(key string) string {
	h := sha1.New()
	io.WriteString(h, key)
	io.WriteString(h, wsGUID)
	return base64.StdEncoding.EncodeToString(h.Sum(nil))
}

// reject writes he's status and reason to w and returns he, so a handler can do
// `if _, err := Accept(w, r); err != nil { return }`.
func reject(w http.ResponseWriter, he *HandshakeError) error {
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	if he.Status == http.StatusUpgradeRequired {
		// Advertise what we require (RFC 7231 §6.5.15).
		w.Header().Set("Upgrade", "websocket")
		w.Header().Set("Connection", "Upgrade")
	}
	w.WriteHeader(he.Status)
	io.WriteString(w, he.Reason+"\n")
	return he
}

// Accept performs the server-side WebSocket handshake requiring the
// DefaultSubprotocol (h2ts), hijacks the connection, and returns it as a [*Conn].
//
// Equivalent to [AcceptWithOptions] with a zero [AcceptOptions]: h2ts is echoed
// when the client offered it; a client that doesn't offer h2ts is rejected with
// 400, a non-upgrade request with 426. On any such rejection Accept writes the
// response to w and returns a [*HandshakeError] — the caller should just return.
//
// On success it writes the 101 and returns a ready-to-use [*Conn]; hand it to
// [ServeH2] (typically on a new goroutine so the accepting handler can return).
// The outer server must be plain HTTP/1.1 so the request is hijackable.
func Accept(w http.ResponseWriter, r *http.Request) (*Conn, error) {
	return AcceptWithOptions(w, r, AcceptOptions{})
}

// AcceptWithOptions is [Accept] with an explicit [AcceptOptions] — a custom
// subprotocol selector, implicit-codec acceptance, and/or an Origin allowlist.
func AcceptWithOptions(w http.ResponseWriter, r *http.Request, opts AcceptOptions) (*Conn, error) {
	if !IsUpgradeRequest(r) {
		return nil, reject(w, errNotUpgrade)
	}

	// Opt-in Origin allowlist (CSWSH defence). Off by default (nginx-style): only
	// enforced when the caller configured AllowedOrigins. A missing Origin is
	// rejected when an allowlist is set.
	if opts.AllowedOrigins != nil {
		origin := r.Header.Get("Origin")
		if origin == "" || !containsFold(opts.AllowedOrigins, origin) {
			return nil, reject(w, errForbiddenOrigin)
		}
	}

	offered := OfferedProtocols(r)
	chosen, hasChosen, accept := negotiate(offered, opts.Select, opts.AllowImplicitCodec)
	if !accept {
		return nil, reject(w, errUnsupportedSubprotocol)
	}

	hj, ok := w.(http.Hijacker)
	if !ok {
		// The outer connection can't be hijacked (e.g. it's itself HTTP/2). The
		// h2ts gateway must run over HTTP/1.1.
		return nil, reject(w, &HandshakeError{
			http.StatusInternalServerError,
			"connection does not support hijacking (serve the gateway over HTTP/1.1)",
		})
	}
	conn, brw, err := hj.Hijack()
	if err != nil {
		return nil, err
	}

	// Write the 101 through the hijacked buffered writer (flushing any bytes
	// net/http had buffered first), then flush to the socket. Reads use brw.Reader,
	// which may already hold client bytes sent after the handshake (an h2ts client
	// pipelines its HTTP/2 preface immediately) — Conn consumes from it.
	if err := writeSwitchingProtocols(brw, r.Header.Get("Sec-WebSocket-Key"), chosen, hasChosen); err != nil {
		conn.Close()
		return nil, err
	}

	proto := ""
	if hasChosen {
		proto = chosen
	}
	return newConn(conn, brw.Reader, proto), nil
}

// writeSwitchingProtocols emits the 101 handshake response.
func writeSwitchingProtocols(brw *bufio.ReadWriter, key, proto string, hasProto bool) error {
	var b strings.Builder
	b.WriteString("HTTP/1.1 101 Switching Protocols\r\n")
	b.WriteString("Upgrade: websocket\r\n")
	b.WriteString("Connection: Upgrade\r\n")
	b.WriteString("Sec-WebSocket-Accept: ")
	b.WriteString(acceptKey(key))
	b.WriteString("\r\n")
	if hasProto {
		b.WriteString("Sec-WebSocket-Protocol: ")
		b.WriteString(proto)
		b.WriteString("\r\n")
	}
	b.WriteString("\r\n")
	if _, err := brw.WriteString(b.String()); err != nil {
		return err
	}
	return brw.Flush()
}
